build/dart/rules.gni

```diff
@@ -5,7 +5,178 @@
 # This file has rules for making Dart packages and Dart-based Mojo applications.
 # The entrypoint is the dart_pkg rule.
 
+import("//build/compiled_action.gni")
 import("//build/module_args/dart.gni")
+import("//flutter/common/config.gni")
+import("//third_party/dart/build/dart/dart_action.gni")
+
+# Creates a dart kernel (dill) file suitalbe for use with gen_snapshot.
+#
+# Invoker must supply dart_main and package_config. Invoker may optionally
+# supply aot as a boolean and product as a boolean.
+template("dart_snapshot") {
+  assert(defined(invoker.main_dart), "main_dart is a required parameter.")
+  assert(defined(invoker.sources), "sources is a required parameter.")
+  assert(defined(invoker.package_config),
+         "package_config is a required parameter.")
+
+  kernel_target = "_${target_name}_kernel"
+  snapshot_target = "_${target_name}_snapshot"
+  is_aot =
```

Comment:
Nit: Perhaps add `assert(!is_fuchsia)` here as this would be invalid on that target.

@richkadel: FYI, you may run into such assumptions while working with engine GN rules for Dart. We can patch them later though. The issue is that the assumption of AOT being tied to a specific runtime mode does not hold on Fuchsia.

Reply:
Fuchsia has its own template(s) for this. 

Perhaps we could add it as an arg and default it to this, and let fuchsia specify the arg differently?

Reply:
`flutter_runtime_mode` can also be `jit_release`. It might simplify things for Fuchsia to use that to signal that a release build is using JIT rather than AOT (if it isn't like that already.)

Reply:
For now I'm just moving this to `//flutter/testing/scenario_app/runtime_mode.gni`.

Reply:
Errr... wrong comment thread - I moved it from `common.gni`. I'll add the assert.